cibuildwheel/__main__.py

```diff
@@ -259,6 +262,10 @@ def main() -> None:
     else:
         assert_never(platform)
 
+    if not identifiers:
```

Comment:
2 things:

1. So, the main difference with #179 is that @Czaki put the detection in the `build` function, and raised an exception there. I think I prefer this approach (checking in `__main__`), since we don't really care about doing this twice?
   -  @Czaki's approach does fix the whole "marker for empty build" issue, since `build` gets intercepted.
2. Why do this after you print the preamble and after calling `build`? If we're catching errors in `__main__.py`, they should probable be thrown before we give the impression that `cibuildwheel` started running successfully?

Reply:
> Why do this after you print the preamble

Because that information could potentially be useful in helping debug what went wrong. The preamble is also not build-specific, it doesn't get repeated N times for N builds.

> since we don't really care about doing this twice

You mean three times? Once per OS file?

Reply:
Though it's not as important now that we have the build selector. However, basic info like what it's running on, etc. still seem quite useful.

Reply:
Hmmm, myeah, maybe. We don't do that for the other errors, though; but you could argue that the other errors are a bit "simpler".

> > since we don't really care about doing this twice
> 
> You mean three times? Once per OS file?

That wasn't clear. But I meant determining the selected build identifiers twice. It's now done to check they're not empty, then again in `{linux,macos,windows}.build()` to actually loop over the different builds.

So I meant we don't really care about the performance of doing this twice, in this case (not like it's going to be measurable), and there's hardly any code duplication.

Reply:
I'm happy to print it out before if that's better; including the build specifier helps quite a bit.

Reply:
I think it's good to print this after the preamble, printing the options (and configuration warnings) might be useful information. But after calling `platform-specific.build` is strange to me... as this check is designed to flag a misconfiguration. Also, just a few lines up we create the output_dir ready to receive wheels, we should be doing this check and failing before that I think.